### PR TITLE
Space key doesn't work when editing filters #1176

### LINKED ISSUE #1167: F1 key doesn't work when the focus is in a filter box
F1 should work even when the focus is in a filter box. Even better if pressing F1 while in the filter box goes directly to the 'filters' section of the user docs.

### LINKED ISSUE #1176: Space key doesn't work when editing filters
When editing a filter, the space key doesn't work when the cursor is at certain locations of the filter.

### PR DESCRIPTION
**Pull request by dariusf**
Fixes #1176.

### CODE REVIEW

**Comment by dariusf** [comment]
@HansNewbie could you help to review? Thanks!

**Comment by HansNewbie** [comment]
For the space testing, can we have some boundary tests as well? Maybe test adding space at start, end, and middle of a word that has no space or start and end with space? For example,
`count` -> `_count`
`count` -> `count_`
`count` -> `cou_nt`
`_count` -> `__count` -> `__count_` -> `__count__`

**Comment by dariusf** [comment]
Added the comment.

Added the last one -- insertion of spaces when the field contains trailing spaces. That's significant because of the legacy code removed. The rest are all in the same class of boundary case and should be covered already, since spaces are not treated specially now.

File: src/test/java/guitests/FilterTextFieldTest.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    @Test
+    public void inputsTest() {
+        FilterTextField field = getFirstPanelField();
+        testCompletions(field);
+        testSpaces(field);
+    }
+
+//    @Test
```
Comment (HansNewbie) [inline src/test/java/guitests/FilterTextFieldTest.java]: Is this meant to be commented? The `cancelTest` as well.

Comment (dariusf) [inline src/test/java/guitests/FilterTextFieldTest.java]: No, good thing you caught it. :sweat:

File: src/main/java/ui/components/FilterTextField.java [hunk 2]
```diff
@@ -87,22 +88,39 @@ private void setup() {
...
-        String word = getCurrentWord() + e.getCharacter();
+    /**
+     * Completion is only started when there's effectively nothing (only whitespace)
+     * after the caret, or if there is selected text (indicating either that we are in
+     * the midst of completion, or that the user does not want the selected text and is
+     * typing to replace it)
+     * @return true if completion should be started
+     */
+    private boolean shouldStartCompletion() {
+        return getCharAfterCaret().trim().isEmpty() || !getSelectedText().isEmpty();
```
Comment (HansNewbie) [inline src/main/java/ui/components/FilterTextField.java]: I need confirmation on this behaviour. What if the user select text in the middle of the filter? Should completion be performed? Say `assignee:me milestone:current` and the selected text is `ignee` of `assignee`.

Comment (dariusf) [inline src/main/java/ui/components/FilterTextField.java]: With `ignee` selected, if you type `i`, the result is `assi[gnee]:me` (selection represented with square brackets). Typing anything else (say, `a`) results in `assa:me`. Completion is performed, but only if the current word + currently-typed letter is a prefix of some keyword.

The main reason for this change was to [not have completion cancelled as the user continues typing](https://github.com/HubTurbo/HubTurbo/blob/1167-space-key/src/test/java/guitests/FilterTextFieldTest.java#L22-L32). The other scenario of typing with some text already selected is a side effect.

Comment (HansNewbie) [inline src/main/java/ui/components/FilterTextField.java]: Yep, I noticed about the completion toggling. Since this is an intended behaviour, sure. This is fine.

File: src/main/java/filter/expression/QualifierType.java [hunk 3]
```diff
@@ -50,7 +50,7 @@
         defaultCompletions.removeAll(Arrays.asList("m", "as", "au", "s"));
 
         defaultCompletions.addAll(Arrays.asList("closed", "open", "issue", "pr",
-            "pullrequest", "read", "unread",  "merged", "unmerged", "comments",
+            "pullrequest", "read", "unread", "me", "merged", "unmerged", "comments",
```
Comment (HansNewbie) [inline src/main/java/filter/expression/QualifierType.java]: Does the alias `me` for `assignee` exist? I was using it in my previous comment for simplicity.

Comment (dariusf) [inline src/main/java/filter/expression/QualifierType.java]: Ugh, just checked and it seems to have been removed at some point. It used to use your login name. Never mind then.

**Comment by HansNewbie** [comment]
Ok seems good.
